File: packages/attachments/src/ports.ts
import type { AttachmentId, AttachmentRef, FileInput, NoteId } from "@notes-app/domain";

/**
 * Insert / delete on `attachments_refs`. The production binding wraps
 * PostgREST under the user's JWT (RLS confines the row set); tests
 * pass an in-memory fake.
 */
export interface AttachmentsRepo {
  /**
   * Persist the row that the pipeline just minted. The implementation
   * is responsible for stamping `owner_id` from the user JWT (via RLS)
   * and returning the canonical row.
   */
  insert(input: NewAttachmentRefInput): Promise<AttachmentRef>;

  /** Delete by id. Idempotent — missing rows count as success. */
  delete(id: AttachmentId): Promise<void>;

  /**
   * Paged scan over all rows for the validator. The Edge Function calls
   * this with the service-role key; client code never does.
   */
  listForValidation(cursor: string | null, limit: number): Promise<ValidationPage>;

  /** Record a dangling-ref entry. Used by the validator. */
  recordDangling(entry: DanglingEntry): Promise<void>;
}

export interface NewAttachmentRefInput {
  readonly noteId: NoteId;
  readonly externalId: string;
  readonly mimeType: string;
  readonly sizeBytes: number;
  readonly displayName: string;
  readonly thumbnailId?: string;
}

/**
 * Note: owner_id is intentionally NOT in [NewAttachmentRefInput]. The
 * repo binding stamps it from the auth.uid() of the user's JWT
 * server-side (via RLS), so the client can't get it wrong. See the
 * stamping trigger pattern used by note_tags in M04.
 */

export interface ValidationPage {
  /** Plain-shape rows so the validator stays JSON-driven. */
  readonly rows: readonly ValidationRow[];
  /** Opaque cursor for the next page; null when caller has reached the end. */
  readonly nextCursor: string | null;
}

export interface ValidationRow {
  readonly id: string;
  readonly external_id: string;
  readonly thumbnail_id: string | null;
  readonly owner_id: string;
}

export interface DanglingEntry {
  readonly attachmentId: string;
  readonly ownerId: string;
  readonly externalId: string;
  readonly missingThumbnail: boolean;
  readonly missingOriginal: boolean;
}

/**
 * Platform port for thumbnail generation. Web wires HTMLCanvasElement /
 * OffscreenCanvas; Android wires android.graphics.Bitmap. Tests pass a
 * function that returns canned bytes.
 */
export interface ThumbnailGenerator {
  /**
   * Generate a thumbnail of [input]. The implementation MUST honour
   * `ThumbnailSpec.maxLongestEdgePx` and emit JPEG at quality 80; the
   * pipeline trusts the produced bytes verbatim.
   */
  generate(input: FileInput): Promise<FileInput>;
}

File: packages/android/attachments-android/src/main/kotlin/app/notes/attachments/Ports.kt
package app.notes.attachments

import app.notes.domain.AttachmentId
import app.notes.domain.AttachmentRef
import app.notes.domain.FileInput
import app.notes.domain.NoteId
import app.notes.domain.Progress
import kotlinx.coroutines.flow.Flow

/**
 * The attachment pipeline port. Kotlin twin of the TS
 * `AttachmentPipeline`. Lives in this module (not in `core-domain`)
 * because `Flow` would pull `kotlinx-coroutines-core` into pure-types
 * domain — we keep that boundary clean.
 */
interface AttachmentPipeline {
    suspend fun attach(noteId: NoteId, file: FileInput): AttachmentRef
    suspend fun resolveUrl(ref: AttachmentRef): String
    suspend fun remove(ref: AttachmentRef)
    fun uploadProgress(ref: AttachmentRef): Flow<Progress>
}

interface AttachmentsRepo {
    suspend fun insert(input: NewAttachmentRefInput): AttachmentRef
    suspend fun delete(id: AttachmentId)
    suspend fun listForValidation(cursor: String?, limit: Int): ValidationPage
    suspend fun recordDangling(entry: DanglingEntry)
}

/**
 * Note: owner_id is intentionally NOT in [NewAttachmentRefInput]. The
 * repo binding stamps it from auth.uid() of the user's JWT server-side
 * (RLS), so the client can't get it wrong.
 */
data class NewAttachmentRefInput(
    val noteId: NoteId,
    val externalId: String,
    val mimeType: String,
    val sizeBytes: Long,
    val displayName: String,
    val thumbnailId: String? = null,
) {
    init {
        require(externalId.isNotBlank()) { "NewAttachmentRefInput.externalId cannot be empty" }
        require(displayName.isNotBlank()) { "NewAttachmentRefInput.displayName cannot be empty" }
        require(sizeBytes >= 0) { "NewAttachmentRefInput.sizeBytes must be >= 0" }
    }
}

data class ValidationPage(
    val rows: List<ValidationRow>,
    val nextCursor: String?,
)

data class ValidationRow(
    val id: String,
    val externalId: String,
    val thumbnailId: String?,
    val ownerId: String,
)

data class DanglingEntry(
    val attachmentId: String,
    val ownerId: String,
    val externalId: String,
    val missingOriginal: Boolean,
    val missingThumbnail: Boolean,
)

/** Platform port for thumbnail generation. Android wires Bitmap. */
fun interface ThumbnailGenerator {
    suspend fun generate(input: FileInput): FileInput
}
